ogr/ogrsf_frmts/mssqlspatial/ogrmssqlspatialtablelayer.cpp

```diff
@@ -1562,13 +1562,15 @@ OGRErr OGRMSSQLSpatialTableLayer::CreateFeatureBCP(OGRFeature *poFeature)
             poSession->CommitTransaction(); /* commit creating the table */
 
         /* Get the column definitions for this table. */
-        if (poFeatureDefn)
+        if (poFeatureDefn && bLayerDefnNeedsRefresh)
         {
-            /* need to re-create layer defn */
-            poFeatureDefn->Release();
-            poFeatureDefn = nullptr;
+            GetLayerDefn();
+            bLayerDefnNeedsRefresh = false;
+        }
```

Comment:
I'm not really understanding this game of why we need to refresh the layer definition. Did you experiment with *not* refreshing it ? although I suspect there must be a good reason for what it has been put in place.

But if that's needed, given that BCP mode can be ended in EndCopy(), I suspect we should also refresh the layer defn in a similar way there too. So probably that bLayerDefnNeedsRefresh should be initialized to false in the .h file, not tested at line 1565 and just set to true before calling GetLayerDefn() at line 1567.

And a similar thing to be done in EndCopy()

Reply:
> I'm not really understanding this game of why we need to refresh the layer definition. Did you experiment with _not_ refreshing it ? although I suspect there must be a good reason for what it has been put in place.
> 

It took me a while to find why, the reason is because of the side effect of GetLayerDefn that calls BuildFeatureDefn() during the layer creation, when fields were not yet added. 

> But if that's needed, given that BCP mode can be ended in EndCopy(), I suspect we should also refresh the layer defn in a similar way there too. So probably that bLayerDefnNeedsRefresh should be initialized to false in the .h file, not tested at line 1565 and just set to true before calling GetLayerDefn() at line 1567.
> 

Done.

> And a similar thing to be done in EndCopy()

That shouldn't be necessary.